File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lastfm/util/AppUtils.java
package com.wa2c.android.medoly.plugin.action.lastfm.util;

import android.content.ContentResolver;
import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.support.v4.content.FileProvider;

import com.google.gson.Gson;
import com.wa2c.android.medoly.plugin.action.lastfm.BuildConfig;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;


/**
 * App utilities.
 */
public class AppUtils {

    /**
     * Show message.
     * @param context context.
     * @param text message.
     */
    public static void showToast(Context context, String text) {
        ToastReceiver.showToast(context, text);
    }

    /**
     * Show message.
     * @param context context
     * @param stringId resource id.
     */
    public static void showToast(Context context, int stringId) {
        ToastReceiver.showToast(context, stringId);
    }

    /**
     * Save object to shared preference.
     * @param context context
     * @param prefKey preference key.
     * @param saveObject save object.
     * @return succeeded / failed
     */
    public static boolean saveObject(Context context, String prefKey,  Object saveObject) {
        try {
            Gson gson = new Gson();
            String json = gson.toJson(saveObject);

            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
            pref.edit().putString(prefKey, json).apply();
            return true;
        } catch (Exception e) {
            Logger.e(e);
            return false;
        }
    }

    /**
     * Load object from shared preference.
     * @param context Context.
     * @param prefKey Preference key.
     * @param clazz Object class.
     * @return Loaded object. null as failed.
     */
    public static <T> T loadObject(Context context, String prefKey, Class<T> clazz) {
        try {
            SharedPreferences pref = PreferenceManager.getDefaultSharedPreferences(context);
            String json = pref.getString(prefKey, "");

            Gson gson = new Gson();

            return gson.fromJson(json, clazz);
        } catch (Exception e) {
            Logger.e(e);
            return null;
        }
    }



    private static final String SHARED_DIR_NAME = "download";
    private static final String PROVIDER_AUTHORITIES = BuildConfig.APPLICATION_ID + ".fileprovider";

    /**
     * Download URI data.
     * @param context A context.
     * @param downloadUrl Download URI.
     * @return Shared URI.
     */
    public static Uri downloadUrl(Context context, String downloadUrl) {
        InputStream inputStream = null;
        FileOutputStream outputStream = null;
        Uri providerUri;
        try {
            final URL url = new URL(downloadUrl);
            if (ContentResolver.SCHEME_FILE.equals(url.getProtocol()) ||
                    ContentResolver.SCHEME_CONTENT.equals(url.getProtocol()))
            {
                return Uri.parse(downloadUrl);
            }

            // Folder
            final File sharedDir = new File(context.getFilesDir(), SHARED_DIR_NAME);
            if (!sharedDir.exists()) {
                // Create a folder
                sharedDir.mkdir();
            } else {
                // Delete all files
                File[] files = sharedDir.listFiles();
                for (File f : files) {
                    if (f.isFile())
                        f.delete();
                }
            }

            // InputStream
            HttpURLConnection con = (HttpURLConnection) url.openConnection();
            con.setRequestMethod("GET");
            con.connect();
            final int status = con.getResponseCode();
            if (status != HttpURLConnection.HTTP_OK) {
                return null;
            }
            inputStream = con.getInputStream();

            // OutputStream
            String[] pathElements = url.getPath().split("/");
            String saveFileName = pathElements[pathElements.length - 1];
            final File sharedFile = new File(sharedDir, saveFileName);
            outputStream = new FileOutputStream(sharedFile);

            // Stream copy
            byte[] buffer = new byte[16384];
            while (true) {
                int len = inputStream.read(buffer);
                if (len < 0) {
                    break;
                }
                outputStream.write(buffer, 0, len);
            }
            outputStream.flush();

            providerUri = FileProvider.getUriForFile(context, PROVIDER_AUTHORITIES, sharedFile);
        } catch (Exception e) {
            return null;
        } finally {
            if (inputStream != null)
                try { inputStream.close(); } catch (IOException e) { e.printStackTrace(); }
            if (outputStream != null)
                try { outputStream.close(); } catch (IOException e) { e.printStackTrace(); }
        }

        // URI
        return providerUri;
    }
}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lastfm/util/AppUtils.kt
package com.wa2c.android.medoly.plugin.action.lastfm.util

import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.core.content.FileProvider
import com.wa2c.android.medoly.library.ExtraData
import com.wa2c.android.medoly.library.MediaPluginIntent
import com.wa2c.android.medoly.library.PropertyData
import com.wa2c.android.medoly.plugin.action.lastfm.BuildConfig
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.File
import java.net.HttpURLConnection
import java.net.URL


/**
 * App utilities.
 */
object AppUtils {

    private const val SHARED_DIR_NAME = "download"
    private const val PROVIDER_AUTHORITIES = BuildConfig.APPLICATION_ID + ".fileprovider"

    /**
     * Download URI data.
     * @param context A context.
     * @param downloadUrl Download URI.
     * @return Shared URI.
     */
    fun downloadUrl(context: Context, downloadUrl: String): Uri? {
        try {
            val url = URL(downloadUrl)
            if (ContentResolver.SCHEME_FILE == url.protocol || ContentResolver.SCHEME_CONTENT == url.protocol) {
                return Uri.parse(downloadUrl)
            }

            // Folder
            val sharedDir = File(context.filesDir, SHARED_DIR_NAME)
            if (!sharedDir.exists()) {
                // Create a folder
                sharedDir.mkdir()
            } else {
                // Delete all files
                sharedDir.listFiles()?.filter { it.isFile }?.forEach { it.delete() }
            }

            // InputStream
            val con = url.openConnection() as HttpURLConnection
            con.requestMethod = "GET"
            con.connect()
            val status = con.responseCode
            if (status != HttpURLConnection.HTTP_OK) {
                return null
            }

            val pathElements = url.path.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val saveFileName = pathElements[pathElements.size - 1]
            val sharedFile = File(sharedDir, saveFileName)

            BufferedInputStream(con.inputStream).use { input ->
                BufferedOutputStream(sharedFile.outputStream()).use { output ->
                    val buffer = ByteArray(16384)
                    while (true) {
                        val len = input.read(buffer)
                        if (len < 0) {
                            break
                        }
                        output.write(buffer, 0, len)
                    }
                    output.flush()
                    return FileProvider.getUriForFile(context, PROVIDER_AUTHORITIES, sharedFile)
                }
            }
        } catch (e: Exception) {
            return null
        }
    }

    /**
     * Send result.
     * @param context A context.
     * @param pluginIntent A plugin intent.
     * @param resultProperty A result property data.
     * @param resultExtra A result extra data.
     */
    fun sendResult(context: Context, pluginIntent: MediaPluginIntent, resultProperty: PropertyData? = null, resultExtra: ExtraData? = null) {
        context.sendBroadcast(pluginIntent.createResultIntent(resultProperty, resultExtra))
    }

    /**
     * Start web page.
     * @param uri Open url.
     */
    fun Context.startPage(uri: Uri) {
        val launchIntent = Intent(Intent.ACTION_VIEW, uri)
        startActivity(launchIntent)
    }
}
